### PR TITLE
Apache HTTP Client 5 implementation

### PR DESCRIPTION
**Pull request by nck-mlcnv**
This PR implements the Apache HTTP Client 5 for the query executions. This should improve the performance of Iguana.
Also changes following things:

- Fix logging configuration
- Cache HTTP requests if queries are cached and also preload them before starting the stresstest (#223)
- HTTP requests send chunked data, if the queries are not cached
- If `parseResults` is turned off, received response bodies from HTTP requests won't be hashed anymore, to improve performance

### CODE REVIEW

**Review by bigerl — CHANGES_REQUESTED** [review]
Thx for the PR. This looks overall sound but could use some more documentation here and there. 
Find my specific comments and remarks below. 
Please also review the usage of the apache http client5 with respect to refactoring, simplifying the code, extracting classes to separate files and documentation. Currently, the usage is a but hard to follow.

File: src/main/java/org/aksw/iguana/cc/utils/http/StreamEntityProducer.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        return -1;
+    }
+
+    @Override
+    public String getContentType() {
+        return null;
+    }
+
+    @Override
+    public String getContentEncoding() {
```
[the remark this answers is not in the document]
Comment (nck-mlcnv) [reply src/main/java/org/aksw/iguana/cc/utils/http/StreamEntityProducer.java:92]: The `Content-Encoding` header is meant for compressions, if I understood it correctly ([link](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding?retiredLocale=de)).
The character encoding for the request body is defined with the `Content-Type` header. In the case of `application/sparql-query` it is already defined to be UTF-8 encoded ([link](https://www.w3.org/TR/sparql11-protocol/#query-via-post-direct)).

File: src/main/java/org/aksw/iguana/cc/utils/http/StreamEntityProducer.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        if (content != null) {
+            return content.remaining();
+        }
+        if (currentStream != null) {
+            try {
+                return currentStream.available();
+            } catch (IOException e) {
+                logger.error("Failed to get available bytes from input stream", e);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (nck-mlcnv) [reply src/main/java/org/aksw/iguana/cc/utils/http/StreamEntityProducer.java:118]: If `content` is not null, it means the whole entity data has been read into the buffer from a stream that was taken from the stream supplier and that the content will be sent non-chunked.
In this case, the remaining bytes in the buffer are returned.

Otherwise, the data is sent in chunks. If there is currently a stream open, from which the data is being read from, the available bytes from that stream are returned.

I added these as comments.

File: src/main/java/org/aksw/iguana/cc/utils/http/StreamEntityProducer.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+            return content.limit();
+        }
+
+        // if the content length is unknown (chunked request), return -1
+        return -1;
+    }
+
+    @Override
+    public String getContentType() {
+        return null;
```
[the remark this answers is not in the document]
Comment (nck-mlcnv) [reply src/main/java/org/aksw/iguana/cc/utils/http/StreamEntityProducer.java]: The `Content-Type` header was already set at another location, but I changed it now anyway.

File: src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java [hunk 4]
```diff
@@ -180,7 +180,13 @@ public QueryStringWrapper getNextQuery(QuerySelector querySelector) throws IOExc
 
     public QueryStreamWrapper getNextQueryStream(QuerySelector querySelector) throws IOException {
         final var queryIndex = querySelector.getNextIndex();
-        return new QueryStreamWrapper(queryIndex, this.queryList.getQueryStream(queryIndex));
+        return new QueryStreamWrapper(queryIndex, config.caching(), () -> {
+            try {
+                return this.queryList.getQueryStream(queryIndex);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
```
[the remark this answers is not in the document]
Comment (nck-mlcnv) [reply src/main/java/org/aksw/iguana/cc/query/handler/QueryHandler.java:188]: I added an additional check now in the `executeQuery` method to check, what methods should be executed. If the query handler cached its queries the worker retrieves the request directly now from the factory through an additional method.

File: src/main/java/org/aksw/iguana/cc/worker/impl/SPARQLProtocolWorker.java [hunk 5]
```diff
@@ -307,128 +283,211 @@ private ExecutionStats executeQuery(Duration timeout, boolean discardOnFailure)
...
+            private final StreamingXXHash64 hasher = hasherFactory.newStreamingHash64(0);
+            private long responseSize = 0; // will be used if parseResults is false
+            private long responseEnd = 0;  // time in nanos
+
+            @Override
+            public void releaseResources() {} // nothing to release
+
+            @Override
+            protected int capacityIncrement() {
+                return Integer.MAX_VALUE; // get as much data in as possible
```
Comment (bigerl) [inline src/main/java/org/aksw/iguana/cc/worker/impl/SPARQLProtocolWorker.java:337]: I am not completely sure where exactly this size ends up being used. I didn't find the documentation of AbstractBinDataConsumer very helpful for this. For safety reasons I would suggest to use: `Integer.MAX_VALUE - 8` instead. That way it is guaranteed that it would not throw creating a java byte array with the capacity.

**Review by bigerl — CHANGES_REQUESTED** [review]
looks good. thx. just this one change and we should be good to go code-wise.

**Review by bigerl — APPROVED** [review]
